File: lib/admin/features/moderation/services/admin_moderation_service.dart
import 'package:cloud_firestore/cloud_firestore.dart' as fs;

import '../../../../core/models/enums.dart';
import '../../../../core/services/firebase_service.dart';
import '../../../../features/ratings/models/rating_model.dart';
import '../../../../features/reports_safety/models/report_model.dart';
import '../../../core/admin_audit_service.dart';

class ReportPage {
  final List<Report> reports;
  final fs.DocumentSnapshot<Map<String, dynamic>>? lastDoc;
  final bool hasMore;
  const ReportPage({
    required this.reports,
    required this.lastDoc,
    required this.hasMore,
  });
}

class RatingPage {
  final List<Rating> ratings;
  final fs.DocumentSnapshot<Map<String, dynamic>>? lastDoc;
  final bool hasMore;
  const RatingPage({
    required this.ratings,
    required this.lastDoc,
    required this.hasMore,
  });
}

/// Trust & safety moderation: reports, safety alerts, and rating takedowns.
/// All three are admin-write-only in `firestore.rules` — the app can only
/// create a report/alert or leave a rating; resolving, annotating, or
/// deleting is exclusively this screen's job.
class AdminModerationService {
  AdminModerationService._();
  static final AdminModerationService instance = AdminModerationService._();

  fs.CollectionReference<Map<String, dynamic>> get _reportsCol =>
      FirebaseService.instance.firestore.collection('reports');
  fs.CollectionReference<Map<String, dynamic>> get _safetyAlertsCol =>
      FirebaseService.instance.firestore.collection('safetyAlerts');
  fs.CollectionReference<Map<String, dynamic>> get _ratingsCol =>
      FirebaseService.instance.firestore.collection('ratings');

  // ── Reports ──────────────────────────────────────────────────────

  Future<ReportPage> fetchReportPage({
    ReportStatus? status,
    fs.DocumentSnapshot<Map<String, dynamic>>? startAfter,
    int limit = 25,
  }) async {
    fs.Query<Map<String, dynamic>> query = _reportsCol;
    if (status != null) {
      query = query.where('status', isEqualTo: status.name);
    }
    query = query.orderBy('createdAt', descending: true);
    if (startAfter != null) query = query.startAfterDocument(startAfter);
    final snap = await query.limit(limit + 1).get();
    final docs = snap.docs;
    final hasMore = docs.length > limit;
    final pageDocs = hasMore ? docs.sublist(0, limit) : docs;
    return ReportPage(
      reports: pageDocs.map((d) => Report.fromMap(d.data())).toList(),
      lastDoc: pageDocs.isEmpty ? null : pageDocs.last,
      hasMore: hasMore,
    );
  }

  Future<Report?> fetchReport(String id) async {
    final doc = await _reportsCol.doc(id).get();
    if (!doc.exists || doc.data() == null) return null;
    return Report.fromMap(doc.data()!);
  }

  Future<void> resolveReport(
    Report report, {
    required ReportStatus status,
    required String adminNote,
    required String adminId,
  }) async {
    await _reportsCol.doc(report.id).update({
      'status': status.name,
      'adminNote': adminNote,
      'resolvedByAdminId': adminId,
      'updatedAt': fs.Timestamp.fromDate(DateTime.now()),
    });
    await AdminAuditService.instance.log(
      action: 'report.resolve',
      targetType: 'report',
      targetId: report.id,
      details: {'status': status.name, 'note': adminNote},
    );
  }

  // ── Safety alerts ────────────────────────────────────────────────

  /// Live — an unresolved safety alert is urgent enough that a paginated
  /// pull-to-refresh model is the wrong fit.
  Stream<List<SafetyAlert>> watchSafetyAlerts({bool? resolved}) {
    fs.Query<Map<String, dynamic>> query = _safetyAlertsCol;
    if (resolved != null) {
      query = query
          .where('resolved', isEqualTo: resolved)
          .orderBy('raisedAt', descending: true);
    } else {
      query = query.orderBy('raisedAt', descending: true);
    }
    return query
        .limit(200)
        .snapshots()
        .map(
          (snap) =>
              snap.docs.map((d) => SafetyAlert.fromMap(d.data())).toList(),
        );
  }

  Future<void> resolveSafetyAlert(
    SafetyAlert alert, {
    required String adminId,
  }) async {
    await _safetyAlertsCol.doc(alert.id).update({
      'resolved': true,
      'resolvedByAdminId': adminId,
      'resolvedAt': fs.Timestamp.fromDate(DateTime.now()),
    });
    await AdminAuditService.instance.log(
      action: 'safety_alert.resolve',
      targetType: 'safetyAlert',
      targetId: alert.id,
      details: {},
    );
  }

  // ── Ratings ──────────────────────────────────────────────────────

  Future<RatingPage> fetchRatingPage({
    int? maxStars,
    fs.DocumentSnapshot<Map<String, dynamic>>? startAfter,
    int limit = 25,
  }) async {
    fs.Query<Map<String, dynamic>> query = _ratingsCol;
    if (maxStars != null) {
      // `stars <= maxStars` combined with `orderBy(createdAt)` would need a
      // composite index this app doesn't declare, and reviews worth
      // moderating are a small slice of all ratings — order by stars
      // instead, which needs none beyond the automatic single-field index.
      query = query
          .where('stars', isLessThanOrEqualTo: maxStars)
          .orderBy('stars')
          .orderBy('createdAt', descending: true);
    } else {
      query = query.orderBy('createdAt', descending: true);
    }
    if (startAfter != null) query = query.startAfterDocument(startAfter);
    final snap = await query.limit(limit + 1).get();
    final docs = snap.docs;
    final hasMore = docs.length > limit;
    final pageDocs = hasMore ? docs.sublist(0, limit) : docs;
    return RatingPage(
      ratings: pageDocs.map((d) => Rating.fromMap(d.data())).toList(),
      lastDoc: pageDocs.isEmpty ? null : pageDocs.last,
      hasMore: hasMore,
    );
  }

  Future<void> deleteRating(Rating rating) async {
    await _ratingsCol.doc(rating.id).delete();
    await AdminAuditService.instance.log(
      action: 'rating.delete',
      targetType: 'rating',
      targetId: rating.id,
      details: {'jobId': rating.jobId, 'stars': rating.stars},
    );
  }
}

File: admin-web/src/services/adminModerationService.ts
import {
  collection, doc, getDoc, getDocs, deleteDoc, updateDoc,
  query, where, orderBy, startAfter, limit as fsLimit, Timestamp,
  type QueryDocumentSnapshot, type DocumentData,
} from 'firebase/firestore'
import { db } from '@/lib/firebase'
import { logAuditEvent } from './adminAuditService'
import { mapReport, mapSafetyAlert, mapRating, type Report, type Rating, type SafetyAlert } from '@/types/models'
import type { ReportStatus } from '@/types/enums'

const reportsCol = () => collection(db, 'reports')
const safetyAlertsCol = () => collection(db, 'safetyAlerts')
const ratingsCol = () => collection(db, 'ratings')

// ── Reports ──────────────────────────────────────────────────────────────

export interface ReportPage {
  reports: Report[]
  lastDoc: QueryDocumentSnapshot<DocumentData> | null
  hasMore: boolean
}

export async function fetchReportPage(opts: {
  status?: ReportStatus
  startAfterDoc?: QueryDocumentSnapshot<DocumentData>
  pageLimit?: number
}): Promise<ReportPage> {
  const limitValue = opts.pageLimit ?? 25
  const constraints = []
  if (opts.status) constraints.push(where('status', '==', opts.status))
  constraints.push(orderBy('createdAt', 'desc'))
  if (opts.startAfterDoc) constraints.push(startAfter(opts.startAfterDoc))
  constraints.push(fsLimit(limitValue + 1))

  const snap = await getDocs(query(reportsCol(), ...constraints))
  const docs = snap.docs
  const hasMore = docs.length > limitValue
  const pageDocs = hasMore ? docs.slice(0, limitValue) : docs
  return {
    reports: pageDocs.map((d) => mapReport(d.id, d.data())),
    lastDoc: pageDocs.length ? pageDocs[pageDocs.length - 1] : null,
    hasMore,
  }
}

export async function fetchReport(id: string): Promise<Report | null> {
  const snap = await getDoc(doc(reportsCol(), id))
  return snap.exists() ? mapReport(snap.id, snap.data()) : null
}

export async function resolveReport(
  report: Report,
  status: ReportStatus,
  adminNote: string,
  adminId: string,
): Promise<void> {
  await updateDoc(doc(reportsCol(), report.id), {
    status,
    adminNote,
    resolvedByAdminId: adminId,
    updatedAt: Timestamp.fromDate(new Date()),
  })
  // Lift the job's hasOpenReport flag once the report is settled, mirroring
  // the mobile app's (now-removed) JobService.clearReportFlag — otherwise a
  // resolved/rejected report leaves the job permanently flagged.
  if (report.jobId && (status === 'resolved' || status === 'rejected')) {
    await updateDoc(doc(collection(db, 'jobs'), report.jobId), {
      hasOpenReport: false,
    })
  }
  await logAuditEvent({
    action: 'report.resolve',
    targetType: 'report',
    targetId: report.id,
    details: { status, note: adminNote },
  })
}

// ── Safety alerts ────────────────────────────────────────────────────────

/** Live — an unresolved safety alert is urgent enough that pull-to-refresh pagination is the wrong fit. */
export function safetyAlertsQuery(resolved?: boolean) {
  const constraints = []
  if (resolved !== undefined) {
    constraints.push(where('resolved', '==', resolved), orderBy('raisedAt', 'desc'))
  } else {
    constraints.push(orderBy('raisedAt', 'desc'))
  }
  constraints.push(fsLimit(200))
  return query(safetyAlertsCol(), ...constraints)
}

export { mapSafetyAlert }

export async function fetchSafetyAlert(id: string): Promise<SafetyAlert | null> {
  const snap = await getDoc(doc(safetyAlertsCol(), id))
  return snap.exists() ? mapSafetyAlert(snap.id, snap.data()) : null
}

export async function resolveSafetyAlert(alertId: string, adminId: string): Promise<void> {
  await updateDoc(doc(safetyAlertsCol(), alertId), {
    resolved: true,
    resolvedByAdminId: adminId,
    resolvedAt: Timestamp.fromDate(new Date()),
  })
  await logAuditEvent({
    action: 'safety_alert.resolve',
    targetType: 'safetyAlert',
    targetId: alertId,
    details: {},
  })
}

// ── Ratings ──────────────────────────────────────────────────────────────

export interface RatingPage {
  ratings: Rating[]
  lastDoc: QueryDocumentSnapshot<DocumentData> | null
  hasMore: boolean
}

/** stars <= maxStars combined with orderBy(createdAt) would need an undeclared composite index — order by stars instead, needing only the automatic single-field index. */
export async function fetchRatingPage(opts: {
  maxStars?: number
  startAfterDoc?: QueryDocumentSnapshot<DocumentData>
  pageLimit?: number
}): Promise<RatingPage> {
  const limitValue = opts.pageLimit ?? 25
  const constraints = []
  if (opts.maxStars !== undefined) {
    constraints.push(where('stars', '<=', opts.maxStars), orderBy('stars'), orderBy('createdAt', 'desc'))
  } else {
    constraints.push(orderBy('createdAt', 'desc'))
  }
  if (opts.startAfterDoc) constraints.push(startAfter(opts.startAfterDoc))
  constraints.push(fsLimit(limitValue + 1))

  const snap = await getDocs(query(ratingsCol(), ...constraints))
  const docs = snap.docs
  const hasMore = docs.length > limitValue
  const pageDocs = hasMore ? docs.slice(0, limitValue) : docs
  return {
    ratings: pageDocs.map((d) => mapRating(d.id, d.data())),
    lastDoc: pageDocs.length ? pageDocs[pageDocs.length - 1] : null,
    hasMore,
  }
}

export async function deleteRating(rating: Pick<Rating, 'id' | 'jobId' | 'stars'>): Promise<void> {
  await deleteDoc(doc(ratingsCol(), rating.id))
  await logAuditEvent({
    action: 'rating.delete',
    targetType: 'rating',
    targetId: rating.id,
    details: { jobId: rating.jobId, stars: rating.stars },
  })
}
